articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+This article will serve as an introduction to some of the important things that an ABI defines looking at how some current ABIs defines them.
+
+# Primitive types
+Primitive types are generally anything that the computer can operate on. Some examples are:
+- Integers: `char`, `short`, `int`, `long`, `unsigned long`, etc in C and `i8`, `i16`, `i32`, `isize`, `usize`, etc in Rust
+- Floating-points: `float`, `double` in C and `f32`, `f64` in Rust (There are more sizes of floats, but we won't talk about them)
+- Booleans: `bool` in most languages.
+- Pointers: `T*` (where `T` is a type) in C and `*const T` or `*mut T` in Rust to name a few.
+- SIMD Vectors: This is SSE or AVX.
+- and more
+
+The primitive types can be grouped into [categories](https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf#section.3.2), what I'll call type-kinds (this is an unofficial term from [this article](https://gankra.github.io/blah/rust-layouts-and-abis/#abi) that I liked and will also use in this article). Most commonly these are:
+- Integers: This includes all integer primitive types as well as booleans.
+- Floating point: This only includes the above-mentioned floating-point primitives (Note, I am simplifying the scope of this type-kind).
+- Pointer: For C/C++ ABIs this contains pointer and reference types, however, in Rust this type-kind [does not exist](https://gankra.github.io/blah/rust-layouts-and-abis/#abi) and instead pointer and reference types are in the integers type-kind.
+- Vector: This is SSE or AVX datatype (we won't talk about this type-kind).
+- Aggregate: General type-kind for structs and classes (we will talk about this type-kind in the [structs](#structs) section).
+
+Sidebar: Booleans in [C](https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf#section.3.2), [C++](https://isocpp.org/files/papers/n4296.pdf#section.3.9), and [Rust](https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#bool) all *only* have acceptable values of 0 and 1 (even though the size is 1 byte or 8 bits). This is because, as shown in these examples, [C](https://repl.it/@ZackJorquera/C-bools-only-0-or-1), [C++](https://repl.it/@ZackJorquera/Cpp-bools-only-0-or-1), and [Rust](https://repl.it/@ZackJorquera/Rust-bools-only-0-or-1) only look at the 0th bit to determine the truth value. This means the while a value like `4`, `0b0100`, is not zero it will still be false as a bool (technically it is undefined behavior). However, in an if statement and when casting there is an [implicit conversion](https://isocpp.org/files/papers/n4296.pdf#section.4.12) that will take any non-zero value and make it have the value 1 if it isn't already the `bool` type.

```

Comment:
like `4`, `0b0100`, is not zero ->  like `4`, `0b0100`, are not zero

Reply:
i changed it to like `4` (`0b0100` in binary) is not